rust/src/arrow/linalg.rs

```diff
@@ -0,0 +1,258 @@
+        let n = self.value_length();
+        let values = self.values();
+        let flatten_values: &Float32Array = as_primitive_array(values.as_ref());
+
+        // Solving: A = U * Sigma * Vt
+        //
+        // TODO: Lapacke requires a mutable reference of `A`.
+        // How can we get mutable reference without copying data?
+        let mut a = transpose(flatten_values.values(), n as usize);
+        // f32 array builder for matrix U.
+        let mut u = vec![0.0; (m * m) as usize];
+        // f32 array builder for matrix V_T
+        let mut vt = vec![0.0; (n * n) as usize];
+        let mut sigma = vec![0.0; n as usize];
+
+        let mut work = vec![0_f32; 1];
+        // Length of the workspace
+        let lwork: i32 = -1;
+        // SGESVG return value.
+        let mut info: i32 = -1;
+
+        // Query the optimal workspace size, will be stored in `work[0]`.
+        unsafe {
+            sgesvd(
+                b'A',
+                b'A',
+                m,
+                n,
+                a.as_mut_slice(),
+                m,
+                sigma.as_mut_slice(),
+                u.as_mut_slice(),
+                m,
+                vt.as_mut_slice(),
+                m,
+                work.as_mut_slice(),
+                lwork,
+                &mut info,
+            );
+        }
+        if info > 0 {
+            println!("Failed to compute sgesvd");
+        }
+
+        let lwork = work[0] as i32;
+        let mut work = vec![0_f32; lwork as usize];
+        unsafe {
+            sgesvd(
+                b'A',
+                b'A',
+                m,
+                n,
+                a.as_mut_slice(),
+                m,
+                sigma.as_mut_slice(),
+                u.as_mut_slice(),
+                m,
+                vt.as_mut_slice(),
+                n,
+                work.as_mut_slice(),
+                lwork,
+                &mut info,
+            );
+        }
+        if info != 0 {
+            return Err(Error::Arrow("Failed to compute SVD".to_string()));
+        }
+
+        let u_values = Float32Array::from_iter_values(transpose(&u, m as usize));
+        let u = Self::try_new(&u_values, m)?;
+        let vt_values = Float32Array::from_iter_values(transpose(&vt, n as usize));
+        let vt = Self::try_new(&vt_values, n)?;
+        let sigma = Float32Array::from_iter_values(sigma);
+        Ok((u, sigma, vt))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use approx::assert_relative_eq;
+    use arrow::datatypes::Float32Type;
+
+    use super::*;
+
+    #[test]
+    fn test_svd() {
+        let values = Float32Array::from_iter_values([
+            8.79, 9.93, 9.83, 5.45, 3.16, 6.11, 6.91, 5.04, -0.27, 7.98, -9.15, -7.93, 4.86, 4.85,
+            3.01, 9.57, 1.64, 8.83, 0.74, 5.80, -3.49, 4.02, 9.80, 10.00, 4.27, 9.84, 0.15, -8.99,
+            -6.02, -5.31,
+        ]);
+        let a = FixedSizeListArray::try_new(&values, 5).unwrap();
```

Comment:
Can you add code comments on the dimensions here?

So `a` is supposed to be a [6x5] matrix right?

Reply:
Yes , will do